src/annotate/onclass/script.py

```diff
@@ -0,0 +1,209 @@
+    return cl_id_to_name, name_to_cl_id
+
+def predict_input_data(model: OnClassModel,
+                       input_matrix: np.array,
+                       input_modality: ad.AnnData,
+                       id_to_name: dict,
+                       obs_prediction: str,
+                       obs_probability: str) -> ad.AnnData:
+    """
+    Predict cell types for input data and save results to Anndata obj.
+    
+    Parameters
+    ----------
+    model : OnClassModel
+        The OnClass model.
+    input_matrix : np.array
+        The input data matrix.
+    input_modality : ad.AnnData
+        The input data Anndata object.
+    id_to_name : dict
+        Dictionary mapping cell ontology IDs to cell type names.
+    obs_prediction : str
+        The obs key for the predicted cell type.
+    obs_probability : str
+        The obs key for the predicted cell type probability.
+        
+    Returns
+    -------
+    ad.AnnData
+        The input data Anndata object with the predicted cell types saved in obs.
+    """
+    corr_test_feature = model.ProcessTestFeature(
+        test_feature=input_matrix,
+        test_genes=input_modality.var_names,
+        log_transform=False,
+    )
+    onclass_pred = model.Predict(corr_test_feature, use_normalize=False, refine=True, unseen_ratio=-1.0)
+    pred_label = [model.i2co[ind] for ind in onclass_pred[2]]
+    pred_cell_type_label = [id_to_name[id] for id in pred_label]
+    
+    input_modality.obs[obs_prediction] = pred_cell_type_label
+    input_modality.obs[obs_probability] = np.max(onclass_pred[1], axis=1) / onclass_pred[1].sum(1)
+    return input_modality
+
+def main():
+    logger.info("Reading input data")
+    input_mudata = mu.read_h5mu(par["input"])
+    input_modality = input_mudata.mod[par["modality"]].copy()
+
+    id_to_name, name_to_id = map_celltype_to_ontology_id(par["cl_obo_file"])
+    obs_predictions = par["output_obs_predictions"] if par["output_obs_predictions"] else [f"{target}_pred" for target in par["reference_obs_targets"]]
+    obs_probabilities = par["output_obs_probability"] if par["output_obs_probability"] else [f"{target}_prob" for target in par["reference_obs_targets"]]
+    
+    if par["input_layer"]:
+        input_matrix = input_modality.layers[par["input_layer"]].toarray()
+    else:
+        input_matrix = input_modality.X.toarray()
+        
+    if par["reference"]:
+        logger.info("Reading reference data")
+        model = OnClassModel(cell_type_nlp_emb_file=par["cl_nlp_emb_file"],
+                             cell_type_network_file=par["cl_ontology_file"])
+        
+        reference_mudata = mu.read_h5mu(par["reference"])
+        reference_modality = reference_mudata.mod[par["modality"]].copy()
+
+        reference_modality.var["gene_symbol"] = list(reference_modality.var.index)
+        reference_modality.var.index = [re.sub("\\.[0-9]+$", "", s) for s in reference_modality.var["ensemblid"]]
```

Comment:
This will error when 'ensemblid' is not an existing var column, can it be adjusted to how it was implemented in celltypist?